src/Steps/Release/MergeUpRelease.php

```diff
@@ -0,0 +1,181 @@
+
+        $this->mergeRecursive($output, $this->getReleasePlan(), $input->getOption('delete-release-branch') ?? false);
+
+        $this->log($output, "All modules merged up");
+    }
+
+    /**
+     * Merge a library and its children up to their major branches
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     * @param bool $deleteBranch Whether to delete the release branch after merging
+     */
+    protected function mergeRecursive(OutputInterface $output, LibraryRelease $releasePlanNode, bool $deleteBranch)
+    {
+        $currentBranch = $releasePlanNode->getLibrary()->getBranch();
+
+        // Skip upgrade-only modules and anything not on an actual branch (e.g. using a pre-existing tag)
+        if ($releasePlanNode->getLibrary()->isUpgradeOnly() || !$currentBranch) {
+            return;
+        }
+        $currentLibrary = $releasePlanNode->getLibrary()->getName();
+
+        // Skip if it's not a proper release branch.
+        $currentBranchIsRelease = preg_match(
+            '/' . preg_quote(RewriteReleaseBranches::RELEASE_SUFFIX) . '$/',
+            $currentBranch
+        );
+        if (!$currentBranchIsRelease) {
+            $this->log(
+                $output,
+                "Warning: Current branch '$currentBranch' is not a release branch. "
+                    . "Skipping $currentLibrary and its children.",
+                'bg=yellow'
+            );
+            return;
+        }
+
+        // Do the merge up recursively
+        foreach ($releasePlanNode->getItems() as $item) {
+            $this->mergeRecursive($output, $item, $deleteBranch);
+        }
+        $this->mergeLibrary($output, $releasePlanNode);
+
+        // Optionally delete the branch from the remote
+        if ($deleteBranch) {
+            $this->log(
+                $output,
+                "Deleting branch <info>$currentBranch</info> from remote for <info>{$currentLibrary}</info>"
+            );
+            $releasePlanNode->getLibrary()->getRepository()->run('push', ['--delete', $currentBranch]);
+        }
+    }
+
+    /**
+     * Performs a merge up for a single library
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     */
+    protected function mergeLibrary(OutputInterface $output, LibraryRelease $releasePlanNode)
+    {
+        $library = $releasePlanNode->getLibrary();
+        $releaseBranch = $library->getBranch();
+        $minorBranch = str_replace(RewriteReleaseBranches::RELEASE_SUFFIX, '', $releaseBranch);
+        $majorBranch = $releasePlanNode->getVersion()->getMajor();
+
+        // Merge up to minor branch, then to major branch
+        $this->mergeUp($output, $releasePlanNode, $releaseBranch, $minorBranch, false);
+        $this->mergeUp($output, $releasePlanNode, $minorBranch, $majorBranch, true);
```

Comment:
I'm not sure how much to worry about automatically merging up JS at this point. This could be a real blocker to doing this via cow.

What we should be doing here, is if there were JS src changes we'd be checking out silverstripe/admin at the corresponding version, running yarn install there, and running yarn build in the current module

We do something similar in CI when running js tests - https://github.com/silverstripe/gha-run-tests/blob/1.0/action.yml#L184

Reply:
If there are JS changes to merge up, cow will bail when the dist bundle merge conflict innevitably occurs, and the release manager can then do those steps manually.
I think this is probably safer than automating js merge-ups - consider for example if there's a merge conflict in some file other than the bundle.js in the same merge - we'd have to detect that separately.

Reply:
Yeah I guess this is fine. If we mess up the js on merge-up, it'll only affect dev branches and show in CI.  The actual tagged release won't ship with merge conflicts